Cw3 fixed multisig

ethanfrey opened the pull request:
Closes #8 

Status:

* Init completed
* Handle logic done (minus events/data on result)
* Query logic done

TODO:

- [x] Write unit tests
- [x] Return info on HandleResult (data, logs)
- [x] Write proper README

ethanfrey commented:
@maurolacy this is a super rough draft, but if you want to take a quick look and let me know if I forgot any error cases or such, please do so.

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+    };
+
+    // get next id
+    let id = next_id(&mut deps.storage)?;
```
I guess concurrency checks / contention is handled at the framework / blockchain level.

ethanfrey replied:
Yeah, we are gauranteed single-threaded no race conditions.

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+
+    let mut prop = proposal_read(&deps.storage).load(&proposal_id.to_be_bytes())?;
+    if prop.status != Status::Open {
+        return Err(StdError::generic_err("Proposal is not open for voting"));
```
"Proposal is not open for voting" -> "Proposal is not open for execution" /  "Proposal is not open"?

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+    // we enforce this for now, but maybe it doesn't make sense... since no one can *vote*
+    // after expiration, if it is expired but there were enough yes votes, it means it passed
+    // in the period. Do we enforce execution then as well?
```
> Do we enforce execution then as well?

I think so.

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+        .iter()
+        .any(|x| *x == prop.status)
+    {
+        return Err(StdError::generic_err("Cannot closed completed proposals"));
```
closed -> close.

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+    // ensure it did not pass (think about the above... passed and expired should be EITHER closeable or executable)
+    if prop.yes_weight >= prop.required_weight {
+        return Err(StdError::generic_err("Already passed, try to execute it"));
```
This is consistent with allowing execution of (passed) expired proposals.

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    // set it to failed
+    prop.status = Status::Failed;
```
This is a personal preference, but I don't like the name `Failed`, as it makes me think there was an error or something. Maybe `DidntPass`, or `Rejected`?

Also, `Executed` can be maybe be renamed to `Passed`? Though in that case I like the naming more, because it indicates the associated messages have been dispatched.

ethanfrey replied:
I renamed them. Also added a pass status (for passed, but not executed). We currently do not display these all, but I will try to improve the info in query to do better.

ethanfrey replied:
Updated in 025e717

maurolacy commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+
+    // TODO: using max as default here, also enforce max
+    let cfg = config_read(&deps.storage).load()?;
+    let expires = latest.unwrap_or(cfg.max_voting_period);
```
No need to enforce that `latest` (if it is set) is less than or equal than `cfg.max_voting_period`?

maurolacy replied:
Also, if and when `earliest` is not ignored, some consistency checks (`earliest` < `latest` <= `max_voting_period`) may be a good idea.

ethanfrey replied:
Definitely, that is the meaning of the TODO - to set a max, and I abused it as default.

maurolacy reviewed (commented):
Logic looks good to me.

ethanfrey commented:
Thanks for the feedback. I will update the spec in some places, code in others

ethanfrey commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+};
+use cw2::set_contract_version;
+
+use crate::msg::{HandleMsg, InitMsg, QueryMsg};
```
Yes, those are good for clients. We may want to add additional methods there. I guess I could alias now, but I wanted to make it easier to add more (like cw20-base)

ethanfrey commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+    set_contract_version(&mut deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let total_weight = msg.voters.iter().map(|v| v.weight).sum();
+    let cfg = Config {
+        required_weight: msg.required_weight,
```
Yup, this will be the simple contract. I plan more complex ones - dynamic, different thresholds, etc. This should be simple and solid.

ethanfrey commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+    config(&mut deps.storage).save(&cfg)?;
+
+    // add all voters
+    let mut voters = voter_weight(&mut deps.storage);
```
I was ending up with lots of variables and methods named `voters` and didn't have a better name. I will update, thanks.

ethanfrey commented on contracts/cw3-fixed-multisig/src/contract.rs:
```diff
@@ -0,0 +1,267 @@
+
+    // add the first yes vote from voter
+    let ballot = Ballot {
+        weight,
```
To display it a bit nicer in queries. But yeah, it is a dup.

ethanfrey commented:
@maurolacy I think I finished all logic and updated it with your feedback

When you have time, can you start writing some unit tests (also for the logic I added in cw0)? Just make a branch and push the commits when you finish them. I will merge them in and add some more tests myself possibly. Thanks

orkunkl commented on contracts/cw3-fixed-multisig/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+library = []
+
+[dependencies]
+cosmwasm-std = { version = "0.10.1", features = ["iterator"] }
```
Added task migration to issue https://github.com/CosmWasm/cosmwasm-plus/issues/96

maurolacy replied:
Hi, better to do that in another MR in my opinion.

ethanfrey replied:
Yes, let us merge this first (and the nft contract) before migrating to 0.11.

For contract obligations, I need https://github.com/CosmWasm/cosmwasm-plus/pull/97 and would like this PR to be merged (0.10.x compatible) before end of month, and published as 0.2.2. Once those are merged, let's get the 0.11 branch going and prepare a 0.3.0 release.

orkunkl commented on contracts/cw3-fixed-multisig/src/state.rs:
```diff
@@ -0,0 +1,111 @@
+    bucket_read(PREFIX_PROPOSAL, storage)
+}
+
+pub fn next_id<S: Storage>(storage: &mut S) -> StdResult<u64> {
```
[storage/sequence](https://github.com/CosmWasm/cosmwasm/blob/master/packages/storage/src/sequence.rs) does exactly this logic.

ethanfrey replied:
I didn't know that. And I wrote it :facepalm:

ethanfrey commented:
I just gotta fix up the readme, then this is ready for merge. Last reviews welcome
